### ISSUE TITLE
Controller classes with the same name are not generated even if routes are different

### ISSUE THREAD

**Issue by ajwhitehead88** [domaindrivendev/Swashbuckle#766]
If a controller has the same class name then it appears that the generated documents have neither showing for some reason.

For example:

Having `FirstController` in Project A with a route of `/a/first` and `FirstController` in Project B with a `RoutePrefix` of `/b/first` then neither will output in the documentation.

Having `FirstController` in Project A with a route of `/a/first` and `SecondController` in Project B with a `RoutePrefix` of `/b/first` then both will output in the documentation.

This does not make a difference even if changing the grouping to use the `RoutePrefix` using the `config.GroupActionsBy(KeySelector)` configuration setting.

**Comment by domaindrivendev** [comment]
Are you by any chance using a custom controller selector to route requests to the appropriate namespaced controller? If so, this is likely the root cause of your issue.

Swashbuckle relies heavily on the built-in metadata layer - ApiExplorer. If you're using a custom controller selector, then it's very likely that your implementation is breaking ApiExplorer and hence Swashbuckle. This has come up several times in the past (e.g. #144).

It's worth noting that the WebApi team advise against customizing the controller selector for this reason because, among other issues, it breaks the descriptions generated by ApiExplorer. See the following for more info:

http://blogs.msdn.com/b/webdev/archive/2014/08/22/customizing-web-api-controller-selector.aspx

To get this working correctly with SB, you'll have to start by implementing your controller selection in a way that doesn't break ApiExplorer.

**Comment by ajwhitehead88** [comment]
This is without using a custom controller selector, no.

I attempted a work around by implementing a custom controller selector but it came to nothing in the end really so dropped it and just prefixed the problem controller's name to get things working.

I'm wondering if the controller selector is updated to a "non-standard" one that isn't fully compatible when calling `config.MapHttpAttributeRoutes()`? That might explain things a little bit I think?

**Comment by hiredgunhouse** [comment]
This almost for sure did not work properly with two controllers with the same name name anyway cause one of the controllers will get cached and be used alwas.  
MS does not support this scenario and that's when you have to use custom controler selector like this one:
https://github.com/WebApiContrib/WebAPIContrib/blob/master/src/WebApiContrib/Selectors/NamespaceHttpControllerSelector.cs which comes from Umbraco (see credits in the file).  
And then the whole ApiExplorer thing breaks and you start to have issues @ajwhitehead88 describes and many others have discovered and describe in other issues that reference this one #144.  
The bottom line is the problem that MS created by ignoring controller namespace and then they ask you not to overwrite the controller selector, yeah thanks a lot MS for this great advise!  
Your best solutions in this case is to detect duplicate controller names and disallow them by i.e. throwing an error during startup up or have a test running in your CI to detect this problem.
If you're using attribute routing you're fine and if not then I'd suggest you create a convention to auto-manually register routes explicitly in a way that your forced-to-be-unique-and-thus-possibly-ugly controller names do not surface to the URL level.
